File: internal/engine/engine_test.go
package engine

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

// Use the test binary itself as a disposable CLI, with no paid model or live board.
func TestAgentProcess(t *testing.T) {
	if os.Getenv("TRASHPOSTER_TEST_AGENT") != "test-helper-enabled" {
		return
	}
	mode := ""
	for i, a := range os.Args {
		if a == "--" {
			mode = os.Args[i+1]
			break
		}
	}
	switch mode {
	case "wait":
		time.Sleep(30 * time.Second)
	case "fail":
		os.Exit(2)
	case "empty":
	case "secret":
		fmt.Print("test-secret-bearer")
	case "flood":
		fmt.Print(strings.Repeat("x", 1000100))
	case "context":
		b, _ := io.ReadAll(os.Stdin)
		cwd, _ := os.Getwd()
		_ = json.NewEncoder(os.Stdout).Encode(map[string]string{"cwd": cwd, "prompt": string(b), "token": os.Getenv("SLOPCHAN_TOKEN")})
	case "file":
		_ = os.WriteFile(os.Args[len(os.Args)-1], []byte("the actual post"), 0600)
		fmt.Print("telemetry only")
	case "argv":
		if strings.Contains(os.Args[len(os.Args)-1], "$(touch do-not-create)") {
			fmt.Print("literal argument")
		}
	default:
		fmt.Print("a tiny useful observation")
	}
	os.Exit(0)
}

type boardFixture struct {
	mu     sync.Mutex
	writes int
	paths  []string
	auth   string
	fail   bool
}

func fixture(t *testing.T, mode string, live bool) (*Engine, *boardFixture) {
	t.Helper()
	b := &boardFixture{}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		b.mu.Lock()
		defer b.mu.Unlock()
		b.paths = append(b.paths, r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		if r.Method == "POST" {
			b.writes++
			b.auth = r.Header.Get("Authorization")
			if b.fail {
				http.Error(w, "failed", 500)
				return
			}
			fmt.Fprint(w, `{"post":{"id":42}}`)
			return
		}
		if strings.HasSuffix(r.URL.Path, "/7") {
			fmt.Fprint(w, `{"id":7,"posts":[{"id":7,"text":"the selected discussion"}]}`)
			return
		}
		fmt.Fprint(w, `{"threads":[{"id":7,"full":false,"posts":[{"id":7,"text":"recent context"}]}]}`)
	}))
	t.Cleanup(server.Close)
	directory := t.TempDir()
	e, err := New(directory)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(e.Close)
	executable, _ := os.Executable()
	c := e.config
	c.URL = server.URL
	c.DryRun = !live
	c.Token = "test-secret-bearer"
	c.Posting = "thread"
	c.Agents = []Entry{{ID: "test", Name: "Test agent", Enabled: true, Command: executable, Args: []string{"-test.run=^TestAgentProcess$", "--", mode}, Delivery: "stdin", Output: "stdout", Env: map[string]string{"TRASHPOSTER_TEST_AGENT": "test-helper-enabled"}}}
	c.Spaces = []Entry{{ID: "test", Name: "Test workspace", Path: directory, Enabled: true}}
	c.Selection = Selection{"test", "test", "lurker"}
	if mode == "file" {
		c.Agents[0].Output = "file"
		c.Agents[0].Args = append(c.Agents[0].Args, "{output}")
	}
	saveConfig(t, e, c)
	return e, b
}
func saveConfig(t *testing.T, e *Engine, c Config) {
	t.Helper()
	b, _ := json.Marshal(c)
	if err := e.Save(b); err != nil {
		t.Fatal(err)
	}
}
func waitRun(t *testing.T, e *Engine) Run {
	t.Helper()
	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		s := e.State()
		if s.Running == "" && len(s.Runs) > 0 {
			return s.Runs[0]
		}
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatal("run did not finish")
	return Run{}
}
func launch(t *testing.T, e *Engine) Run {
	t.Helper()
	if err := e.Launch(); err != nil {
		t.Fatal(err)
	}
	return waitRun(t, e)
}
func TestPreviewContext(t *testing.T) {
	t.Setenv("SLOPCHAN_TOKEN", "inherited-secret")
	e, b := fixture(t, "context", false)
	r := launch(t, e)
	if r.Status != "preview" {
		t.Fatalf("%+v", r)
	}
	var output map[string]string
	if err := json.Unmarshal([]byte(r.Text), &output); err != nil {
		t.Fatal(err)
	}
	dir, _ := filepath.EvalSymlinks(e.directory)
	if output["cwd"] != dir || output["token"] != "" || !strings.Contains(output["prompt"], "recent context") || !strings.Contains(output["prompt"], "Internet cryptid") || strings.Contains(output["prompt"], e.config.Token) {
		t.Fatalf("bad prompt context: %+v", output)
	}
	if b.writes != 0 {
		t.Fatal("preview posted")
	}
}
func TestDelivery(t *testing.T) {
	for _, mode := range []string{"stdout", "file", "argv"} {
		t.Run(mode, func(t *testing.T) {
			e, b := fixture(t, mode, true)
			c := e.config
			c.Posting = "reply"
			if mode == "argv" {
				c.Agents[0].Delivery = "argv"
				c.Prompt = "$(touch do-not-create)"
			}
			saveConfig(t, e, c)
			r := launch(t, e)
			if r.Status != "posted" || r.PostID != 42 || r.Thread != 7 || b.writes != 1 || b.auth != "Bearer test-secret-bearer" {
				t.Fatalf("%+v, %+v", r, b)
			}
			if mode == "file" && r.Text != "the actual post" {
				t.Fatal(r.Text)
			}
			if mode == "argv" && r.Text != "literal argument" {
				t.Fatal(r.Text)
			}
			if !strings.Contains(strings.Join(b.paths, ","), "/api/threads/7/posts") {
				t.Fatal(b.paths)
			}
		})
	}
}
func TestFailuresNeverPost(t *testing.T) {
	for _, mode := range []string{"fail", "empty", "secret", "flood"} {
		t.Run(mode, func(t *testing.T) {
			e, b := fixture(t, mode, true)
			r := launch(t, e)
			if r.Status != "failed" || b.writes != 0 || strings.Contains(r.Log, "test-secret-bearer") {
				t.Fatalf("%+v", r)
			}
		})
	}
}
func TestCancelAndLock(t *testing.T) {
	e, b := fixture(t, "wait", true)
	if err := e.Launch(); err != nil {
		t.Fatal(err)
	}
	if err := e.Launch(); err == nil {
		t.Fatal("overlap accepted")
	}
	raw, _ := json.Marshal(e.config)
	if err := e.Save(raw); err == nil {
		t.Fatal("saved during run")
	}
	time.Sleep(100 * time.Millisecond)
	e.Stop()
	r := waitRun(t, e)
	if r.Status != "cancelled" || b.writes != 0 || e.State().NextAt != 0 {
		t.Fatalf("%+v", r)
	}
}
func TestTimeout(t *testing.T) {
	e, b := fixture(t, "wait", true)
	c := e.config
	c.Schedule.Timeout = 5
	saveConfig(t, e, c)
	r := launch(t, e)
	if r.Status != "cancelled" || !strings.Contains(r.Error, "timed out") || b.writes != 0 {
		t.Fatalf("%+v", r)
	}
}
func TestUncertainWriteNeverRetries(t *testing.T) {
	e, b := fixture(t, "stdout", true)
	b.fail = true
	r := launch(t, e)
	if b.writes != 1 || r.Status != "failed" || !strings.Contains(r.Error, "uncertain") {
		t.Fatalf("%+v", r)
	}
}
func TestPersistenceAndToken(t *testing.T) {
	e, _ := fixture(t, "stdout", false)
	launch(t, e)
	s := e.State()
	if s.Config.Token != "" || !s.Config.HasToken {
		t.Fatal("token exposed")
	}
	saveConfig(t, e, s.Config)
	if e.config.Token != "test-secret-bearer" {
		t.Fatal("token lost")
	}
	if err := e.Start(); err != nil {
		t.Fatal(err)
	}
	next, err := New(e.directory)
	if err != nil {
		t.Fatal(err)
	}
	defer next.Close()
	if next.State().Active || len(next.State().Runs) != 1 {
		t.Fatal("bad restart")
	}
	for _, name := range []string{"config.json", "history.json"} {
		info, err := os.Stat(filepath.Join(e.directory, name))
		if err != nil || info.Mode().Perm() != 0600 {
			t.Fatal("storage not private")
		}
	}
	c := s.Config
	c.Token = ""
	raw, _ := json.Marshal(c)
	var fields map[string]any
	_ = json.Unmarshal(raw, &fields)
	fields["token"] = ""
	raw, _ = json.Marshal(fields)
	e.Stop()
	if err = e.Save(raw); err != nil {
		t.Fatal(err)
	}
	if e.State().Config.HasToken {
		t.Fatal("clear failed")
	}
}
func TestSelectionAndClock(t *testing.T) {
	c := Defaults()
	for i := 0; i < 100; i++ {
		d := interval(c.Schedule)
		if d < 180*time.Second || d > 900*time.Second {
			t.Fatal(d)
		}
	}
	c.Schedule.Mode = "fixed"
	if interval(c.Schedule) != 300*time.Second {
		t.Fatal("fixed interval")
	}
	if _, err := choose(c.Agents, "missing", nil); err == nil {
		t.Fatal("invalid selection")
	}
	e, _ := fixture(t, "stdout", false)
	if err := e.Start(); err != nil {
		t.Fatal(err)
	}
	if e.State().NextAt <= time.Now().UnixMilli() {
		t.Fatal("clock not armed")
	}
	launch(t, e)
	if !e.State().Active || e.State().NextAt == 0 {
		t.Fatal("one-shot did not rearm")
	}
	e.Stop()
	if e.State().NextAt != 0 {
		t.Fatal("clock not stopped")
	}
}
func TestValidation(t *testing.T) {
	for _, mutate := range []func(*Config){func(c *Config) { c.Schedule.Min = 0 }, func(c *Config) { c.Selection.Agent = "missing" }, func(c *Config) { c.Agents[0].Args = nil }, func(c *Config) { c.Agents[0].Env = nil }, func(c *Config) { c.URL = "https://user:pass@example.com" }, func(c *Config) { c.Spaces[0].Path = "relative" }} {
		c := Defaults()
		mutate(&c)
		if Validate(c) == nil {
			t.Fatal("invalid config accepted")
		}
	}
}
func TestReader(t *testing.T) {
	e, b := fixture(t, "stdout", false)
	input := strings.NewReader("{bad\n" + `{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2025-06-18"}}` + "\n" + `{"jsonrpc":"2.0","method":"notifications/initialized"}` + "\n" + `{"jsonrpc":"2.0","id":2,"method":"tools/list"}` + "\n" + `{"jsonrpc":"2.0","id":3,"method":"tools/call","params":{"name":"read_thread","arguments":{"id":7}}}` + "\n" + `{"jsonrpc":"2.0","id":4,"method":"tools/call","params":{"name":"publish","arguments":{}}}` + "\n")
	var out strings.Builder
	if err := ServeReader(e.config.URL, input, &out); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(out.String(), "the selected discussion") || !strings.Contains(out.String(), "Only public board reads") || b.writes != 0 {
		t.Fatal(out.String())
	}
	if _, err := readPath("read_post", map[string]any{"id": 1.5}); err == nil {
		t.Fatal("fraction accepted")
	}
	if _, err := readPath("list_threads", map[string]any{"token": "x"}); err == nil {
		t.Fatal("unknown argument accepted")
	}
	args := explorationArgs("codex", []string{"exec"}, e.config.URL)
	if !strings.Contains(strings.Join(args, " "), "--board-reader") {
		t.Fatal(args)
	}
}

func TestCloseCancelsAndPersists(t *testing.T) {
	e, b := fixture(t, "wait", true)
	if err := e.Launch(); err != nil {
		t.Fatal(err)
	}
	time.Sleep(100 * time.Millisecond)
	e.Close()
	s := e.State()
	if s.Running != "" || s.Active || s.Runs[0].Status != "cancelled" || b.writes != 0 {
		t.Fatalf("shutdown left work active: %+v", s.Runs)
	}
	next, err := New(e.directory)
	if err != nil {
		t.Fatal(err)
	}
	defer next.Close()
	if next.State().Runs[0].Status != "cancelled" {
		t.Fatal("shutdown not persisted")
	}
	if err = e.Launch(); err == nil {
		t.Fatal("closed engine accepted work")
	}
}

File: legacy/test/engine.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, rmSync, statSync, writeFileSync, realpathSync } from 'node:fs';
import http from 'node:http';
import { tmpdir } from 'node:os';
import { resolve } from 'node:path';
import { createStore, defaults, validate } from '../src/config.js';
import { Engine, choose, intervalMs, buildPrompt } from '../src/engine.js';
import { createApp } from '../src/server.js';

function fixture(t, { code = 'process.stdin.resume(); process.stdin.on("end", () => console.log("a tiny useful observation"));', live = false, posting = 'thread', file = false } = {}) {
  const directory = mkdtempSync(resolve(tmpdir(), 'trashposter-test-'));
  const store = createStore(directory);
  const script = resolve(directory, 'fake-agent.cjs'); writeFileSync(script, code);
  const config = store.get();
  config.agents = [{ id: 'test', name: 'Test agent', command: process.execPath, args: [script, ...(file ? ['{output}'] : [])], env: {}, delivery: 'stdin', output: file ? 'file' : 'stdout', enabled: true }];
  config.spaces = [{ id: 'test', name: 'Test workspace', path: directory, enabled: true }];
  config.selection = { agent: 'test', space: 'test', personality: 'lurker' };
  config.dryRun = !live; config.token = 'test-secret-bearer'; config.posting = posting;
  store.save(config);
  const requests = [];
  const fakeFetch = async (url, options) => {
    requests.push({ url, ...options });
    if (options.method === 'POST') return Response.json({ post: { id: 42 } }, { status: 201 });
    if (url.endsWith('/api/threads/7')) return Response.json({ id: 7, posts: [{ id: 7, text: 'the selected discussion' }] });
    return Response.json({ threads: [{ id: 7, full: false, posts: [{ id: 7, text: 'recent context' }] }] });
  };
  const engine = new Engine(store, { fetch: fakeFetch });
  t.after(() => { engine.stop(); rmSync(directory, { recursive: true, force: true }); });
  return { directory, store, engine, requests, config };
}
test('validates intervals, missing references and literal command arguments', () => {
  const c = defaults(); c.schedule.min = 0; assert.throws(() => validate(c), /min/);
  c.schedule.min = 1000; assert.throws(() => validate(c), /Minimum/);
  c.schedule.min = 100; c.selection.agent = 'missing'; assert.throws(() => validate(c), /enabled agent/);
  c.selection.agent = 'random'; c.agents[0].args = ['$(echo nope)', '`whoami`']; c.agents[0].output = 'stdout'; assert.deepEqual(validate(c).agents[0].args, c.agents[0].args);
});
test('clock and selector honor fixed, bounded random and enabled pools', () => {
  const s = { mode: 'random', min: 10, max: 20, interval: 42 };
  assert.equal(intervalMs(s, () => 0), 10000); assert.equal(intervalMs(s, () => .9999), 20000);
  assert.equal(intervalMs({ ...s, mode: 'fixed' }), 42000);
  const items = [{ id: 'off', enabled: false }, { id: 'a', enabled: true }, { id: 'b', enabled: true }];
  assert.equal(choose(items, 'random', () => 0).id, 'a'); assert.equal(choose(items, 'b').id, 'b'); assert.throws(() => choose(items, 'off'));
});
test('preview launches inside the chosen folder with core + personality + board context; never posts', async t => {
  const f = fixture(t, { code: 'let p="";process.stdin.on("data",d=>p+=d);process.stdin.on("end",()=>console.log(JSON.stringify({cwd:process.cwd(),prompt:p,token:process.env.SLOPCHAN_TOKEN||null})));' });
  const run = await f.engine.launch(); assert.equal(run.status, 'preview');
  const result = JSON.parse(run.text);
  assert.equal(result.cwd, realpathSync(f.directory)); assert.equal(result.token, null);
  assert.match(result.prompt, /CORE DIRECTION/); assert.match(result.prompt, /Internet cryptid/); assert.match(result.prompt, /recent context/);
  assert.ok(!result.prompt.includes(f.config.token)); assert.equal(f.requests.filter(r => r.method === 'POST').length, 0);
});
test('live run posts exactly one final response with bearer auth', async t => {
  const f = fixture(t, { live: true }); const run = await f.engine.launch();
  assert.equal(run.status, 'posted'); assert.equal(run.postId, 42);
  const posts = f.requests.filter(r => r.method === 'POST'); assert.equal(posts.length, 1);
  assert.equal(posts[0].headers.Authorization, `Bearer ${f.config.token}`); assert.equal(JSON.parse(posts[0].body).text, 'a tiny useful observation');
  assert.ok(posts[0].url.endsWith('/api/threads')); assert.ok(run.postUrl.endsWith('/posts/42'));
});
test('reply mode retrieves full thread and uses reply endpoint', async t => {
  const f = fixture(t, { live: true, posting: 'reply' }); const run = await f.engine.launch();
  assert.equal(run.thread, 7); assert.equal(run.status, 'posted');
  assert.ok(f.requests.some(r => r.method === 'GET' && r.url.endsWith('/api/threads/7')));
  assert.ok(f.requests.some(r => r.method === 'POST' && r.url.endsWith('/api/threads/7/posts')));
});
test('file adapter ignores stdout telemetry and publishes only response file', async t => {
  const f = fixture(t, { file: true, code: 'process.stdin.resume();process.stdin.on("end",()=>{console.log("telemetry, not a post");require("fs").writeFileSync(process.argv[2],"the actual post");});' });
  const run = await f.engine.launch(); assert.equal(run.status, 'preview'); assert.equal(run.text, 'the actual post'); assert.match(run.log, /telemetry/);
});
test('argv delivery treats prompt metacharacters literally', async t => {
  const f = fixture(t, { code: 'console.log(process.argv[2].includes("$(touch do-not-create)") ? "literal argument" : "bad argument");' });
  const c = f.store.get(); c.prompt = '$(touch do-not-create)'; c.agents[0].delivery = 'argv'; f.store.save(c);
  assert.equal((await f.engine.launch()).text, 'literal argument');
});
test('failed or empty agents never post', async t => {
  for (const code of ['process.exit(2)', 'process.stdin.resume()']) {
    const f = fixture(t, { live: true, code }); const run = await f.engine.launch(); assert.equal(run.status, 'failed'); assert.equal(f.requests.filter(r => r.method === 'POST').length, 0);
  }
});
test('stop cancels process and prohibits a subsequent post; concurrent launch rejected', async t => {
  const f = fixture(t, { live: true, code: 'setTimeout(()=>console.log("too late"),30000)' });
  const job = f.engine.launch();
  await assert.rejects(f.engine.launch(), /already running/);
  await new Promise(r => setTimeout(r, 150)); f.engine.stop();
  const run = await job; assert.equal(run.status, 'cancelled'); assert.equal(f.engine.active, false); assert.equal(f.engine.nextAt, null); assert.equal(f.requests.filter(r => r.method === 'POST').length, 0);
});
test('configured secrets are redacted from logs and blocked from posting', async t => {
  const f = fixture(t, { live: true, code: 'console.log("test-secret-bearer");' });
  const run = await f.engine.launch(); assert.equal(run.status, 'failed'); assert.match(run.error, /secret/); assert.ok(!run.log.includes(f.config.token)); assert.equal(run.text, '');
});
test('network write failure is recorded once, never retried', async t => {
  const f = fixture(t, { live: true }); const original = f.engine.fetch; let writes = 0;
  f.engine.fetch = async (url, opts) => { if (opts.method === 'POST') { writes++; throw new Error('Connection dropped'); } return original(url, opts); };
  const run = await f.engine.launch(); assert.equal(writes, 1); assert.equal(run.status, 'failed'); assert.match(run.error, /uncertain/);
});
test('session timeout terminates a stalled agent without publishing', async t => {
  const f = fixture(t, { live: true, code: 'setTimeout(()=>console.log("late"),30000)' });
  const c = f.store.get(); c.schedule.timeout = 5; f.store.save(c);
  const run = await f.engine.launch(); assert.equal(run.status, 'cancelled'); assert.match(run.error, /timed out/); assert.equal(f.requests.filter(r => r.method === 'POST').length, 0);
});
test('periodic transport launches on its clock and arms the next interval after completion', async t => {
  const f = fixture(t); const c = f.store.get(); c.schedule.mode = 'fixed'; c.schedule.interval = 10; f.store.save(c);
  t.mock.timers.enable({ apis: ['setTimeout', 'Date'], now: 100000 });
  f.engine.start(); assert.equal(f.engine.nextAt, 110000); assert.equal(f.engine.runs.length, 0);
  t.mock.timers.tick(9999); assert.equal(f.engine.runs.length, 0);
  t.mock.timers.tick(1); assert.equal(f.engine.runs.length, 1); assert.ok(f.engine.current); assert.equal(f.engine.nextAt, null);
  for (let tries = 0; f.engine.current && tries < 100000; tries++) await new Promise(r => setImmediate(r));
  assert.equal(f.engine.runs[0].status, 'preview'); assert.equal(f.engine.nextAt, 120000);
  f.engine.stop(); t.mock.timers.tick(20000); assert.equal(f.engine.runs.length, 1);
});
test('config and run history persist privately; restart never rearms', async t => {
  const f = fixture(t); await f.engine.launch(); f.engine.start(); assert.ok(f.engine.nextAt);
  const next = new Engine(createStore(f.directory)); assert.equal(next.active, false); assert.equal(next.runs[0].status, 'preview'); assert.equal(next.store.get().token, f.config.token);
  assert.equal(statSync(resolve(f.directory, 'config.json')).mode & 0o777, 0o600);
  assert.equal(statSync(resolve(f.directory, 'history.json')).mode & 0o777, 0o600);
});
test('HTTP API rejects cross-origin/host and missing session; hides token and preserves it on save', async t => {
  const f = fixture(t); const app = createApp({ directory: f.directory });
  await new Promise(r => app.server.listen(0, '127.0.0.1', r)); t.after(() => new Promise(r => app.server.close(r)));
  const base = `http://127.0.0.1:${app.server.address().port}`;
  const state = await (await fetch(base + '/api/state')).json(); assert.equal(state.config.token, undefined); assert.equal(state.config.hasToken, true);
  assert.equal((await fetch(base + '/api/start', { method: 'POST' })).status, 403);
  assert.equal((await fetch(base + '/api/state', { headers: { Origin: 'https://evil.example' } })).status, 403);
  const badHostStatus = await new Promise((resolve, reject) => { http.get(base + '/api/state', { headers: { Host: 'evil.example' } }, res => { res.resume(); resolve(res.statusCode); }).on('error', reject); });
  assert.equal(badHostStatus, 403);
  const response = await fetch(base + '/api/config', { method: 'POST', headers: { 'X-Trashposter-Session': state.session, 'Content-Type': 'application/json' }, body: JSON.stringify(state.config) });
  assert.equal(response.status, 200); assert.equal(app.store.get().token, f.config.token);
  app.engine.start(); const locked = await fetch(base + '/api/config', { method: 'POST', headers: { 'X-Trashposter-Session': state.session }, body: JSON.stringify(state.config) }); assert.equal(locked.status, 400); app.engine.stop();
});
